In [70]:
import numpy as np
import pandas as pd

data2015= pd.read_csv('/home/cristina/Ironhack/WEEK2/2015_happiness.csv')
display((data2015).head())

data2015['Happiness Score']=data2015['Happiness Score']/1000
display(data2015)
data2016['Happiness Score']=data2016['Happiness Score']/1000
display(data2016)

#Más adelante me encontraré con este problema, las unidades del Happiness Score no son las mismas en los tres años

Country          Region  Happiness Rank  Happiness Score  \
0  Switzerland  Western Europe               1           7587.0   
1      Iceland  Western Europe               2           7561.0   
2      Denmark  Western Europe               3           7527.0   
3       Norway  Western Europe               4           7522.0   
4       Canada   North America               5           7427.0   

   Standard Error  Economy (GDP per Capita)   Family  \
0         0.03411                   1.39651  1.34951   
1         0.04884                   1.30232  1.40223   
2         0.03328                   1.32548  1.36058   
3         0.03880                1459.00000  1.33095   
4         0.03553                   1.32629  1.32261   

   Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
0                   0.94143  0.66557                        0.41978   
1                   0.94784  0.62877                        0.14145   
2                   0.87464  0.64938                        0.48357   
3                   0.88521  0.66973                        0.36503   
4                   0.90563  0.63297                        0.32957   

   Generosity  Dystopia Residual  
0     0.29678            2.51738  
1     0.43630            2.70201  
2     0.34139            2.49204  
3     0.34699            2.46531  
4     0.45811            2.45176

Country                           Region  \
0                 Switzerland                   Western Europe   
1                     Iceland                   Western Europe   
2                     Denmark                   Western Europe   
3                      Norway                   Western Europe   
4                      Canada                    North America   
5                     Finland                   Western Europe   
6                 Netherlands                   Western Europe   
7                      Sweden                   Western Europe   
8                 New Zealand        Australia and New Zealand   
9                   Australia        Australia and New Zealand   
10                     Israel  Middle East and Northern Africa   
11                 Costa Rica      Latin America and Caribbean   
12                    Austria                   Western Europe   
13                     Mexico      Latin America and Caribbean   
14              United States                    North America   
15                     Brazil      Latin America and Caribbean   
16                 Luxembourg                   Western Europe   
17                    Ireland                   Western Europe   
18                    Belgium                   Western Europe   
19       United Arab Emirates  Middle East and Northern Africa   
20             United Kingdom                   Western Europe   
21                       Oman  Middle East and Northern Africa   
22                  Venezuela      Latin America and Caribbean   
23                  Singapore                Southeastern Asia   
24                     Panama      Latin America and Caribbean   
25                    Germany                   Western Europe   
26                      Chile      Latin America and Caribbean   
27                      Qatar  Middle East and Northern Africa   
28                     France                   Western Europe   
29                  Argentina      Latin America and Caribbean   
..                        ...                              ...   
128                   Myanmar                Southeastern Asia   
129                   Georgia       Central and Eastern Europe   
130                    Malawi               Sub-Saharan Africa   
131                 Sri Lanka                    Southern Asia   
132                  Cameroon               Sub-Saharan Africa   
133                  Bulgaria       Central and Eastern Europe   
134                     Egypt  Middle East and Northern Africa   
135                     Yemen  Middle East and Northern Africa   
136                    Angola               Sub-Saharan Africa   
137                      Mali               Sub-Saharan Africa   
138       Congo (Brazzaville)               Sub-Saharan Africa   
139                   Comoros               Sub-Saharan Africa   
140                    Uganda               Sub-Saharan Africa   
141                   Senegal               Sub-Saharan Africa   
142                     Gabon               Sub-Saharan Africa   
143                     Niger               Sub-Saharan Africa   
144                  Cambodia                Southeastern Asia   
145                  Tanzania               Sub-Saharan Africa   
146                Madagascar               Sub-Saharan Africa   
147  Central African Republic               Sub-Saharan Africa   
148                      Chad               Sub-Saharan Africa   
149                    Guinea               Sub-Saharan Africa   
150               Ivory Coast               Sub-Saharan Africa   
151              Burkina Faso               Sub-Saharan Africa   
152               Afghanistan                    Southern Asia   
153                    Rwanda               Sub-Saharan Africa   
154                     Benin               Sub-Saharan Africa   
155                     Syria  Middle East and Northern Africa   
156                   Burundi               Sub-Saharan Africa

Country                           Region  Happiness Rank  \
0                 Denmark                   Western Europe               1   
1             Switzerland                   Western Europe               2   
2                 Iceland                   Western Europe               3   
3                  Norway                   Western Europe               4   
4                 Finland                   Western Europe               5   
5                  Canada                    North America               6   
6             Netherlands                   Western Europe               7   
7             New Zealand        Australia and New Zealand               8   
8               Australia        Australia and New Zealand               9   
9                  Sweden                   Western Europe              10   
10                 Israel  Middle East and Northern Africa              11   
11                Austria                   Western Europe              12   
12          United States                    North America              13   
13             Costa Rica      Latin America and Caribbean              14   
14            Puerto Rico      Latin America and Caribbean              15   
15                Germany                   Western Europe              16   
16                 Brazil      Latin America and Caribbean              17   
17                Belgium                   Western Europe              18   
18                Ireland                   Western Europe              19   
19             Luxembourg                   Western Europe              20   
20                 Mexico      Latin America and Caribbean              21   
21              Singapore                Southeastern Asia              22   
22         United Kingdom                   Western Europe              23   
23                  Chile      Latin America and Caribbean              24   
24                 Panama      Latin America and Caribbean              25   
25              Argentina      Latin America and Caribbean              26   
26         Czech Republic       Central and Eastern Europe              27   
27   United Arab Emirates  Middle East and Northern Africa              28   
28                Uruguay      Latin America and Caribbean              29   
29                  Malta                   Western Europe              30   
..                    ...                              ...             ...   
127               Senegal               Sub-Saharan Africa             128   
128              Bulgaria       Central and Eastern Europe             129   
129            Mauritania               Sub-Saharan Africa             130   
130              Zimbabwe               Sub-Saharan Africa             131   
131                Malawi               Sub-Saharan Africa             132   
132                 Sudan               Sub-Saharan Africa             133   
133                 Gabon               Sub-Saharan Africa             134   
134                  Mali               Sub-Saharan Africa             135   
135                 Haiti      Latin America and Caribbean             136   
136              Botswana               Sub-Saharan Africa             137   
137               Comoros               Sub-Saharan Africa             138   
138           Ivory Coast               Sub-Saharan Africa             139   
139              Cambodia                Southeastern Asia             140   
140                Angola               Sub-Saharan Africa             141   
141                 Niger               Sub-Saharan Africa             142   
142           South Sudan               Sub-Saharan Africa             143   
143                  Chad               Sub-Saharan Africa             144   
144          Burkina Faso               Sub-Saharan Africa             145   
145                Uganda               Sub-Saharan Africa             145   
146                 Yemen  Middle East a

#Elijo Kaggle porque no quiero tener problemas a la hora de convertir ficheros, pero todas las bases de datos que hay están bastante limpias (las que yo he chequedo), por lo que decido coger una que sea por lo menos interesante y sobre esa hacer todas las posibles operaciones que se me ocurran. Mi base de datos es un ranking de felicidad por países en función de determinadas variables, como la renta per cápita, la esperanza de vida o el nivel de corrupción del país. Hay 3 años, 2015-2016-2017. Comienzo descargando el primero que previamente he guardado en mi carpeta. Aquí el primer problema,se me olvida una contrabarra al inicio de la ruta y tardo hasta que veo donde está el problema. Una vez solucionado descargo los otros 3 años para empezar a trabajar


In [71]:
data2016= pd.read_csv('/home/cristina/Ironhack/WEEK2/2016_happiness.csv')
data2017= pd.read_csv('/home/cristina/Ironhack/WEEK2/2017_happiness.csv')

In [72]:
display((data2016).head())
display((data2017).head())

Country          Region  Happiness Rank  Happiness Score  \
0      Denmark  Western Europe               1           7526.0   
1  Switzerland  Western Europe               2           7509.0   
2      Iceland  Western Europe               3           7501.0   
3       Norway  Western Europe               4           7498.0   
4      Finland  Western Europe               5           7413.0   

   Lower Confidence Interval  Upper Confidence Interval  \
0                       7.46                    7592.00   
1                    7428.00                       7.59   
2                    7333.00                    7669.00   
3                    7421.00                    7575.00   
4                    7351.00                    7475.00   

   Economy (GDP per Capita)   Family  Health (Life Expectancy)  Freedom  \
0                   1.44178  1.16374                   0.79504  0.57941   
1                   1.52733  1.14524                   0.86303  0.58557   
2                   1.42666  1.18326                   0.86733  0.56624   
3                   1.57744  1.12690                   0.79579  0.59609   
4                   1.40598  1.13464                   0.81091  0.57104   

   Trust (Government Corruption)  Generosity  Dystopia Residual  
0                        0.44453     0.36171            2.73939  
1                        0.41203     0.28083            2.69463  
2                        0.14975     0.47678            2.83137  
3                        0.35776     0.37895            2.66465  
4                        0.41004     0.25492            2.82596

Country  Happiness.Rank  Happiness.Score  Whisker.high  Whisker.low  \
0       Norway               1            7.537      7.594445     7.479556   
1      Denmark               2            7.522      7.581728     7.462272   
2      Iceland               3            7.504      7.622030     7.385970   
3  Switzerland               4            7.494      7.561772     7.426227   
4      Finland               5            7.469      7.527542     7.410458   

   Economy..GDP.per.Capita.    Family  Health..Life.Expectancy.   Freedom  \
0                  1.616463  1.533524                  0.796667  0.635423   
1                  1.482383  1.551122                  0.792566  0.626007   
2                  1.480633  1.610574                  0.833552  0.627163   
3                  1.564980  1.516912                  0.858131  0.620071   
4                  1.443572  1.540247                  0.809158  0.617951   

   Generosity  Trust..Government.Corruption.  Dystopia.Residual  
0    0.362012                       0.315964           2.277027  
1    0.355280                       0.400770           2.313707  
2    0.475540                       0.153527           2.322715  
3    0.290549                       0.367007           2.276716  
4    0.245483                       0.382612           2.430182

#Compruebo que las 3 bases de datos tienen la misma estructura de columnas, para ello saco la longitud de las columnas. El año 2016 tiene una columna adicional,13 en lugar de 12, para ver cual es imprimo las cabeceras por orden afabético para localizarla, pero cada uno de los años tiene un número de registros diferentes, por lo que deduzco que no se ha realizado la encuesta en los mismos países lso tres años, por lo que decido unir las tres tablas pero sólo teniendo en cuenta los países en los que la encuesta se haya realizado los 3 años. 


In [73]:
len(data2015.columns)



12

In [74]:
len(data2016.columns)

13

In [75]:
len(data2017.columns)

12

In [76]:
data2015.columns.sort_values()


Index(['Country', 'Dystopia Residual', 'Economy (GDP per Capita)', 'Family',
       'Freedom', 'Generosity', 'Happiness Rank', 'Happiness Score',
       'Health (Life Expectancy)', 'Region', 'Standard Error',
       'Trust (Government Corruption)'],
      dtype='object')

In [77]:
data2016.columns.sort_values()

Index(['Country', 'Dystopia Residual', 'Economy (GDP per Capita)', 'Family',
       'Freedom', 'Generosity', 'Happiness Rank', 'Happiness Score',
       'Health (Life Expectancy)', 'Lower Confidence Interval', 'Region',
       'Trust (Government Corruption)', 'Upper Confidence Interval'],
      dtype='object')

In [78]:
data2017.columns.sort_values()

Index(['Country', 'Dystopia.Residual', 'Economy..GDP.per.Capita.', 'Family',
       'Freedom', 'Generosity', 'Happiness.Rank', 'Happiness.Score',
       'Health..Life.Expectancy.', 'Trust..Government.Corruption.',
       'Whisker.high', 'Whisker.low'],
      dtype='object')

#En realidad sólo hay una columna diferente, que es Region, hay otras dos en 2016-2017 que han nombrado de manera diferente, entiendo que Whisker.high en 2017 es lo mismo que Upper Confidence Interval en 2016, y  lo mismo sucede con Lower Confidence Interval y Whisker.low. Decido mantener esas columnas porque son similares y  respecto a Region, me parece interesante mantenerla e intentar rellenar con el mismo criterio los otros dos años, ya que lo más probable es que los índices más altos de felicidad se concentren en países europeos (aunque es sabido también que las tasas más altas de suicidio se concentran también en estos países, por lo que sería interesante cruzar el dataframe final con una tabla de este tipo).

**Conclusiones respecto a Columnas**:

    Renombraré las columnas con registros similares pero diferente encabezado
    Mantengo Region e intentaré rellenarlo cuando haga el merge

**Conclusiones respecto a Filas**:
Viendo los países por filas, durante el año 2016 se incluyeron dos países que no están ni en 2015 ni en 2017, a la hora de unir las tablas elimino estos registros para que los datos por año sean comparables
   

In [79]:
len(set(data2015['Country']))


158

In [80]:
len(set(data2016['Country']))

157

In [81]:
len(set(data2017['Country']))

155

#Pensaba renormbrar las columnas, pero en realidad no hace falta porque no las voy a usar, simplemente decido no seleccionarlas cuando haga los dataframes. Creo a continuación 3 dataframes con las mismas columnas para poder hacer un merge

In [82]:
df_2015= pd.DataFrame(data2015,columns =['Country','Region','Happiness Rank', 'Happiness Score','Economy (GDP per Capita)', 'Family','Freedom', 'Generosity', 
       'Health (Life Expectancy)','Trust (Government Corruption)'])

df_2016= pd.DataFrame(data2016,columns =['Country','Region','Happiness Rank', 'Happiness Score','Economy (GDP per Capita)', 'Family','Freedom', 'Generosity', 
       'Health (Life Expectancy)','Trust (Government Corruption)'])

df_2017= pd.DataFrame(data2017,columns =['Country','Region','Happiness Rank', 'Happiness Score','Economy (GDP per Capita)', 'Family','Freedom', 'Generosity', 
       'Health (Life Expectancy)','Trust (Government Corruption)'])


In [83]:
display((df_2017).head())

Country  Region  Happiness Rank  Happiness Score  \
0       Norway     NaN             NaN              NaN   
1      Denmark     NaN             NaN              NaN   
2      Iceland     NaN             NaN              NaN   
3  Switzerland     NaN             NaN              NaN   
4      Finland     NaN             NaN              NaN   

   Economy (GDP per Capita)    Family   Freedom  Generosity  \
0                       NaN  1.533524  0.635423    0.362012   
1                       NaN  1.551122  0.626007    0.355280   
2                       NaN  1.610574  0.627163    0.475540   
3                       NaN  1.516912  0.620071    0.290549   
4                       NaN  1.540247  0.617951    0.245483   

   Health (Life Expectancy)  Trust (Government Corruption)  
0                       NaN                            NaN  
1                       NaN                            NaN  
2                       NaN                            NaN  
3                       NaN                            NaN  
4                       NaN                            NaN

#Al hacer el display de 2017 veo más valores NaN de los que esperaba (esperaba unicamente en Region) veo que las columnas de 2017 realmente no se llamaban exactamente igual por lo que vuelvo atrás para renombrar las columnas de 2017

In [84]:
data2017.columns.sort_values()
display(data2017.columns)

data2017.columns = ['Country', 'Happiness Rank', 'Happiness Score', 'Whisker high',
       'Whisker low', 'Economy (GDP per Capita)', 'Family',
       'Health (Life Expectancy)', 'Freedom', 'Generosity',
       'Trust (Government Corruption)', 'Dystopia Residual']
data2017.columns
display(data2017.head())


Index(['Country', 'Happiness.Rank', 'Happiness.Score', 'Whisker.high',
       'Whisker.low', 'Economy..GDP.per.Capita.', 'Family',
       'Health..Life.Expectancy.', 'Freedom', 'Generosity',
       'Trust..Government.Corruption.', 'Dystopia.Residual'],
      dtype='object')

Country  Happiness Rank  Happiness Score  Whisker high  Whisker low  \
0       Norway               1            7.537      7.594445     7.479556   
1      Denmark               2            7.522      7.581728     7.462272   
2      Iceland               3            7.504      7.622030     7.385970   
3  Switzerland               4            7.494      7.561772     7.426227   
4      Finland               5            7.469      7.527542     7.410458   

   Economy (GDP per Capita)    Family  Health (Life Expectancy)   Freedom  \
0                  1.616463  1.533524                  0.796667  0.635423   
1                  1.482383  1.551122                  0.792566  0.626007   
2                  1.480633  1.610574                  0.833552  0.627163   
3                  1.564980  1.516912                  0.858131  0.620071   
4                  1.443572  1.540247                  0.809158  0.617951   

   Generosity  Trust (Government Corruption)  Dystopia Residual  
0    0.362012                       0.315964           2.277027  
1    0.355280                       0.400770           2.313707  
2    0.475540                       0.153527           2.322715  
3    0.290549                       0.367007           2.276716  
4    0.245483                       0.382612           2.430182

In [85]:
display((df_2017).head())
df_2017= pd.DataFrame(data2017,columns =['Country','Region','Happiness Rank','Happiness Score','Economy (GDP per Capita)', 'Family','Freedom', 'Generosity', 
       'Health (Life Expectancy)','Trust (Government Corruption)'])
df_2017.head()


Country  Region  Happiness Rank  Happiness Score  \
0       Norway     NaN             NaN              NaN   
1      Denmark     NaN             NaN              NaN   
2      Iceland     NaN             NaN              NaN   
3  Switzerland     NaN             NaN              NaN   
4      Finland     NaN             NaN              NaN   

   Economy (GDP per Capita)    Family   Freedom  Generosity  \
0                       NaN  1.533524  0.635423    0.362012   
1                       NaN  1.551122  0.626007    0.355280   
2                       NaN  1.610574  0.627163    0.475540   
3                       NaN  1.516912  0.620071    0.290549   
4                       NaN  1.540247  0.617951    0.245483   

   Health (Life Expectancy)  Trust (Government Corruption)  
0                       NaN                            NaN  
1                       NaN                            NaN  
2                       NaN                            NaN  
3                       NaN                            NaN  
4                       NaN                            NaN

Country  Region  Happiness Rank  Happiness Score  \
0       Norway     NaN               1            7.537   
1      Denmark     NaN               2            7.522   
2      Iceland     NaN               3            7.504   
3  Switzerland     NaN               4            7.494   
4      Finland     NaN               5            7.469   

   Economy (GDP per Capita)    Family   Freedom  Generosity  \
0                  1.616463  1.533524  0.635423    0.362012   
1                  1.482383  1.551122  0.626007    0.355280   
2                  1.480633  1.610574  0.627163    0.475540   
3                  1.564980  1.516912  0.620071    0.290549   
4                  1.443572  1.540247  0.617951    0.245483   

   Health (Life Expectancy)  Trust (Government Corruption)  
0                  0.796667                       0.315964  
1                  0.792566                       0.400770  
2                  0.833552                       0.153527  
3                  0.858131                       0.367007  
4                  0.809158                       0.382612

#Ya están todas las columnas renombradas igual y aparecen los datos que esperabamos(sólo NaN en Region) por lo que puedo hacer el merge

In [86]:
Felicidad = pd.merge(df_2015,df_2016,df_2017, on = 'Country', how='outer')


TypeError: merge() got multiple values for argument 'how'

#El merge me da error porque en realidad estoy uniendo tres años diferentes pero no lo estoy indicando. Tengo que añadir una columna a cada dataframe donde indique a que año corresponde el dato. 

In [87]:
df_2015['year'] = 2015
df_2015.head()

Country          Region  Happiness Rank  Happiness Score  \
0  Switzerland  Western Europe               1            7.587   
1      Iceland  Western Europe               2            7.561   
2      Denmark  Western Europe               3            7.527   
3       Norway  Western Europe               4            7.522   
4       Canada   North America               5            7.427   

   Economy (GDP per Capita)   Family  Freedom  Generosity  \
0                   1.39651  1.34951  0.66557     0.29678   
1                   1.30232  1.40223  0.62877     0.43630   
2                   1.32548  1.36058  0.64938     0.34139   
3                1459.00000  1.33095  0.66973     0.34699   
4                   1.32629  1.32261  0.63297     0.45811   

   Health (Life Expectancy)  Trust (Government Corruption)  year  
0                   0.94143                        0.41978  2015  
1                   0.94784                        0.14145  2015  
2                   0.87464                        0.48357  2015  
3                   0.88521                        0.36503  2015  
4                   0.90563                        0.32957  2015

In [88]:
df_2016['year'] = 2016
df_2017['year'] = 2017


df_2016.head()
df_2017.head()



Country  Region  Happiness Rank  Happiness Score  \
0       Norway     NaN               1            7.537   
1      Denmark     NaN               2            7.522   
2      Iceland     NaN               3            7.504   
3  Switzerland     NaN               4            7.494   
4      Finland     NaN               5            7.469   

   Economy (GDP per Capita)    Family   Freedom  Generosity  \
0                  1.616463  1.533524  0.635423    0.362012   
1                  1.482383  1.551122  0.626007    0.355280   
2                  1.480633  1.610574  0.627163    0.475540   
3                  1.564980  1.516912  0.620071    0.290549   
4                  1.443572  1.540247  0.617951    0.245483   

   Health (Life Expectancy)  Trust (Government Corruption)  year  
0                  0.796667                       0.315964  2017  
1                  0.792566                       0.400770  2017  
2                  0.833552                       0.153527  2017  
3                  0.858131                       0.367007  2017  
4                  0.809158                       0.382612  2017

#Añado las columnas year a los dataframes para poder hacer el merge. Pero no me aparecen los datos, aparece NaN. Hago concat

In [92]:
Felicidad = pd.concat([df_2015, df_2016, df_2017], axis=0)

display(Felicidad.sort_values(by='Happiness Rank', ascending=True))

#A pesar de haber actualizado la tabla Happiness Score dividiendo entre 1000 para que las unidades fueran las mismas la tabla no se refresca



Country                           Region  \
0                 Switzerland                   Western Europe   
0                      Norway                              NaN   
0                     Denmark                   Western Europe   
1                     Iceland                   Western Europe   
1                     Denmark                              NaN   
1                 Switzerland                   Western Europe   
2                     Denmark                   Western Europe   
2                     Iceland                              NaN   
2                     Iceland                   Western Europe   
3                      Norway                   Western Europe   
3                 Switzerland                              NaN   
3                      Norway                   Western Europe   
4                      Canada                    North America   
4                     Finland                              NaN   
4                     Finland                   Western Europe   
5                 Netherlands                              NaN   
5                      Canada                    North America   
5                     Finland                   Western Europe   
6                 Netherlands                   Western Europe   
6                 Netherlands                   Western Europe   
6                      Canada                              NaN   
7                 New Zealand        Australia and New Zealand   
7                      Sweden                   Western Europe   
7                 New Zealand                              NaN   
8                   Australia        Australia and New Zealand   
8                 New Zealand        Australia and New Zealand   
8                      Sweden                              NaN   
9                      Sweden                   Western Europe   
9                   Australia        Australia and New Zealand   
9                   Australia                              NaN   
..                        ...                              ...   
146                     Yemen  Middle East and Northern Africa   
147                   Liberia                              NaN   
147                Madagascar               Sub-Saharan Africa   
147  Central African Republic               Sub-Saharan Africa   
148                  Tanzania               Sub-Saharan Africa   
148                      Chad               Sub-Saharan Africa   
148                    Guinea                              NaN   
149                   Liberia               Sub-Saharan Africa   
149                    Guinea               Sub-Saharan Africa   
149                      Togo                              NaN   
150               Ivory Coast               Sub-Saharan Africa   
150                    Rwanda                              NaN   
150                    Guinea               Sub-Saharan Africa   
151              Burkina Faso               Sub-Saharan Africa   
151                    Rwanda               Sub-Saharan Africa   
151                     Syria                              NaN   
152               Afghanistan                    Southern Asia   
152                     Benin               Sub-Saharan Africa   
152                  Tanzania                              NaN   
153               Afghanistan                    Southern Asia   
153                   Burundi                              NaN   
153                    Rwanda               Sub-Saharan Africa   
154  Central African Republic                              NaN   
154                      Togo               Sub-Saharan Africa   
154                     Benin               Sub-Saharan Africa   
155                     Syria  Middle East and Northern Africa   
155                     Syria  Middle East and Northern Africa   
156                   Burundi               Sub-Saharan Africa   
156                   Burundi               Sub-Saharan Africa

#Me encuentro un problema, los datos de 2015 de Happiness Score no están en la misma base decimal que los del resto de años. Vuelvo a la tabla de 2015 y  divido los datos/1000. Aparentemente también hay un problema en el GDP per cápita de Noruega, 1459 cuando probablemente sea 1.459. Lo dejo para más adelante ver si es un outlier que lo dudo o un error manual. Las instrucciones 95 y 96  son referentes a este problema, comprobaciones, las puedes saltar.


In [95]:
data2015.dtypes

Country                           object
Region                            object
Happiness Rank                     int64
Happiness Score                  float64
Standard Error                   float64
Economy (GDP per Capita)         float64
Family                           float64
Health (Life Expectancy)         float64
Freedom                          float64
Trust (Government Corruption)    float64
Generosity                       float64
Dystopia Residual                float64
dtype: object

In [96]:
display(df_2015)


Country                           Region  \
0                 Switzerland                   Western Europe   
1                     Iceland                   Western Europe   
2                     Denmark                   Western Europe   
3                      Norway                   Western Europe   
4                      Canada                    North America   
5                     Finland                   Western Europe   
6                 Netherlands                   Western Europe   
7                      Sweden                   Western Europe   
8                 New Zealand        Australia and New Zealand   
9                   Australia        Australia and New Zealand   
10                     Israel  Middle East and Northern Africa   
11                 Costa Rica      Latin America and Caribbean   
12                    Austria                   Western Europe   
13                     Mexico      Latin America and Caribbean   
14              United States                    North America   
15                     Brazil      Latin America and Caribbean   
16                 Luxembourg                   Western Europe   
17                    Ireland                   Western Europe   
18                    Belgium                   Western Europe   
19       United Arab Emirates  Middle East and Northern Africa   
20             United Kingdom                   Western Europe   
21                       Oman  Middle East and Northern Africa   
22                  Venezuela      Latin America and Caribbean   
23                  Singapore                Southeastern Asia   
24                     Panama      Latin America and Caribbean   
25                    Germany                   Western Europe   
26                      Chile      Latin America and Caribbean   
27                      Qatar  Middle East and Northern Africa   
28                     France                   Western Europe   
29                  Argentina      Latin America and Caribbean   
..                        ...                              ...   
128                   Myanmar                Southeastern Asia   
129                   Georgia       Central and Eastern Europe   
130                    Malawi               Sub-Saharan Africa   
131                 Sri Lanka                    Southern Asia   
132                  Cameroon               Sub-Saharan Africa   
133                  Bulgaria       Central and Eastern Europe   
134                     Egypt  Middle East and Northern Africa   
135                     Yemen  Middle East and Northern Africa   
136                    Angola               Sub-Saharan Africa   
137                      Mali               Sub-Saharan Africa   
138       Congo (Brazzaville)               Sub-Saharan Africa   
139                   Comoros               Sub-Saharan Africa   
140                    Uganda               Sub-Saharan Africa   
141                   Senegal               Sub-Saharan Africa   
142                     Gabon               Sub-Saharan Africa   
143                     Niger               Sub-Saharan Africa   
144                  Cambodia                Southeastern Asia   
145                  Tanzania               Sub-Saharan Africa   
146                Madagascar               Sub-Saharan Africa   
147  Central African Republic               Sub-Saharan Africa   
148                      Chad               Sub-Saharan Africa   
149                    Guinea               Sub-Saharan Africa   
150               Ivory Coast               Sub-Saharan Africa   
151              Burkina Faso               Sub-Saharan Africa   
152               Afghanistan                    Southern Asia   
153                    Rwanda               Sub-Saharan Africa   
154                     Benin               Sub-Saharan Africa   
155                     Syria  Middle East and Northern Africa   
156                   Burundi               Sub-Saharan Africa

#Aparentemente están los datos limpios a excepción de la columna de Region, que es la única con Nan

In [97]:
Felicidad = pd.concat([df_2015, df_2016, df_2017], axis=0)

display(Felicidad.sort_values(by='Happiness Rank', ascending=True))

Country                           Region  \
0                 Switzerland                   Western Europe   
0                      Norway                              NaN   
0                     Denmark                   Western Europe   
1                     Iceland                   Western Europe   
1                     Denmark                              NaN   
1                 Switzerland                   Western Europe   
2                     Denmark                   Western Europe   
2                     Iceland                              NaN   
2                     Iceland                   Western Europe   
3                      Norway                   Western Europe   
3                 Switzerland                              NaN   
3                      Norway                   Western Europe   
4                      Canada                    North America   
4                     Finland                              NaN   
4                     Finland                   Western Europe   
5                 Netherlands                              NaN   
5                      Canada                    North America   
5                     Finland                   Western Europe   
6                 Netherlands                   Western Europe   
6                 Netherlands                   Western Europe   
6                      Canada                              NaN   
7                 New Zealand        Australia and New Zealand   
7                      Sweden                   Western Europe   
7                 New Zealand                              NaN   
8                   Australia        Australia and New Zealand   
8                 New Zealand        Australia and New Zealand   
8                      Sweden                              NaN   
9                      Sweden                   Western Europe   
9                   Australia        Australia and New Zealand   
9                   Australia                              NaN   
..                        ...                              ...   
146                     Yemen  Middle East and Northern Africa   
147                   Liberia                              NaN   
147                Madagascar               Sub-Saharan Africa   
147  Central African Republic               Sub-Saharan Africa   
148                  Tanzania               Sub-Saharan Africa   
148                      Chad               Sub-Saharan Africa   
148                    Guinea                              NaN   
149                   Liberia               Sub-Saharan Africa   
149                    Guinea               Sub-Saharan Africa   
149                      Togo                              NaN   
150               Ivory Coast               Sub-Saharan Africa   
150                    Rwanda                              NaN   
150                    Guinea               Sub-Saharan Africa   
151              Burkina Faso               Sub-Saharan Africa   
151                    Rwanda               Sub-Saharan Africa   
151                     Syria                              NaN   
152               Afghanistan                    Southern Asia   
152                     Benin               Sub-Saharan Africa   
152                  Tanzania                              NaN   
153               Afghanistan                    Southern Asia   
153                   Burundi                              NaN   
153                    Rwanda               Sub-Saharan Africa   
154  Central African Republic                              NaN   
154                      Togo               Sub-Saharan Africa   
154                     Benin               Sub-Saharan Africa   
155                     Syria  Middle East and Northern Africa   
155                     Syria  Middle East and Northern Africa   
156                   Burundi               Sub-Saharan Africa   
156                   Burundi               Sub-Saharan Africa

Para rellenar los Nan de region lo que hago es intentar crear un dataframe que contenga País-region para poder cruzarlo con el dataframe original

In [ ]:
pais= pd.DataFrame(Felicidad.Country.unique())
display(pais)

region=pd.DataFrame(Felicidad.Region.unique())
display(region)

pais_region= pd.DataFrame(data2016, columns=['Country','Region'])
display(pais_region)

In [120]:
pais_region= pd.DataFrame(data2016, columns=['Country','Region'])
display((pais_region).head())

Country          Region
0      Denmark  Western Europe
1  Switzerland  Western Europe
2      Iceland  Western Europe
3       Norway  Western Europe
4      Finland  Western Europe

No sé cómo continuar, cómo relleno los Nan con la region en función del país? con replace no puedo, lo ideal sería añadir una columna que en  función del país te ponga la región pero no lo consigo


In [133]:
pais2017= pd.DataFrame(data2017.Country.unique())
                       
display(pais2017)
region2017= pd.DataFrame(pais2017,pais_region)
display(region2017)



0
0                      Norway
1                     Denmark
2                     Iceland
3                 Switzerland
4                     Finland
5                 Netherlands
6                      Canada
7                 New Zealand
8                      Sweden
9                   Australia
10                     Israel
11                 Costa Rica
12                    Austria
13              United States
14                    Ireland
15                    Germany
16                    Belgium
17                 Luxembourg
18             United Kingdom
19                      Chile
20       United Arab Emirates
21                     Brazil
22             Czech Republic
23                  Argentina
24                     Mexico
25                  Singapore
26                      Malta
27                    Uruguay
28                  Guatemala
29                     Panama
..                        ...
125          Congo (Kinshasa)
126                      Mali
127               Ivory Coast
128                  Cambodia
129                     Sudan
130                     Ghana
131                   Ukraine
132                    Uganda
133              Burkina Faso
134                     Niger
135                    Malawi
136                      Chad
137                  Zimbabwe
138                   Lesotho
139                    Angola
140               Afghanistan
141                  Botswana
142                     Benin
143                Madagascar
144                     Haiti
145                     Yemen
146               South Sudan
147                   Liberia
148                    Guinea
149                      Togo
150                    Rwanda
151                     Syria
152                  Tanzania
153                   Burundi
154  Central African Republic

[155 rows x 1 columns]

0
(Denmark, Western Europe)                           NaN
(Switzerland, Western Europe)                       NaN
(Iceland, Western Europe)                           NaN
(Norway, Western Europe)                            NaN
(Finland, Western Europe)                           NaN
(Canada, North America)                             NaN
(Netherlands, Western Europe)                       NaN
(New Zealand, Australia and New Zealand)            NaN
(Australia, Australia and New Zealand)              NaN
(Sweden, Western Europe)                            NaN
(Israel, Middle East and Northern Africa)           NaN
(Austria, Western Europe)                           NaN
(United States, North America)                      NaN
(Costa Rica, Latin America and Caribbean)           NaN
(Puerto Rico, Latin America and Caribbean)          NaN
(Germany, Western Europe)                           NaN
(Brazil, Latin America and Caribbean)               NaN
(Belgium, Western Europe)                           NaN
(Ireland, Western Europe)                           NaN
(Luxembourg, Western Europe)                        NaN
(Mexico, Latin America and Caribbean)               NaN
(Singapore, Southeastern Asia)                      NaN
(United Kingdom, Western Europe)                    NaN
(Chile, Latin America and Caribbean)                NaN
(Panama, Latin America and Caribbean)               NaN
(Argentina, Latin America and Caribbean)            NaN
(Czech Republic, Central and Eastern Europe)        NaN
(United Arab Emirates, Middle East and Northern...  NaN
(Uruguay, Latin America and Caribbean)              NaN
(Malta, Western Europe)                             NaN
...                                                 ...
(Senegal, Sub-Saharan Africa)                       NaN
(Bulgaria, Central and Eastern Europe)              NaN
(Mauritania, Sub-Saharan Africa)                    NaN
(Zimbabwe, Sub-Saharan Africa)                      NaN
(Malawi, Sub-Saharan Africa)                        NaN
(Sudan, Sub-Saharan Africa)                         NaN
(Gabon, Sub-Saharan Africa)                         NaN
(Mali, Sub-Saharan Africa)                          NaN
(Haiti, Latin America and Caribbean)                NaN
(Botswana, Sub-Saharan Africa)                      NaN
(Comoros, Sub-Saharan Africa)                       NaN
(Ivory Coast, Sub-Saharan Africa)                   NaN
(Cambodia, Southeastern Asia)                       NaN
(Angola, Sub-Saharan Africa)                        NaN
(Niger, Sub-Saharan Africa)                         NaN
(South Sudan, Sub-Saharan Africa)                   NaN
(Chad, Sub-Saharan Africa)                          NaN
(Burkina Faso, Sub-Saharan Africa)                  NaN
(Uganda, Sub-Saharan Africa)                        NaN
(Yemen, Middle East and Northern Africa)            NaN
(Madagascar, Sub-Saharan Africa)                    NaN
(Tanzania, Sub-Saharan Africa)                      NaN
(Liberia, Sub-Saharan Africa)                       NaN
(Guinea, Sub-Saharan Africa)                        NaN
(Rwanda, Sub-Saharan Africa)                        NaN
(Benin, Sub-Saharan Africa)                         NaN
(Afghanistan, Southern Asia)                        NaN
(Togo, Sub-Saharan Africa)                          NaN
(Syria, Middle East and Northern Africa)            NaN
(Burundi, Sub-Saharan Africa)                       NaN

[157 rows x 1 columns]

In [137]:
df_region = df_2017.map(pais_region)

AttributeError: 'DataFrame' object has no attribute 'map'

#NsdaBelén, que no hay manera,es un buscarv lo que quiero, seguro que se hace en un periquete pero no doy con ello


In [139]:
Felicidad = pd.concat([df_2015, df_2016, df_2017], axis=0)

display(Felicidad.sort_values(by='Happiness Rank', ascending=True))

Country                           Region  \
0                 Switzerland                   Western Europe   
0                      Norway                              NaN   
0                     Denmark                   Western Europe   
1                     Iceland                   Western Europe   
1                     Denmark                              NaN   
1                 Switzerland                   Western Europe   
2                     Denmark                   Western Europe   
2                     Iceland                              NaN   
2                     Iceland                   Western Europe   
3                      Norway                   Western Europe   
3                 Switzerland                              NaN   
3                      Norway                   Western Europe   
4                      Canada                    North America   
4                     Finland                              NaN   
4                     Finland                   Western Europe   
5                 Netherlands                              NaN   
5                      Canada                    North America   
5                     Finland                   Western Europe   
6                 Netherlands                   Western Europe   
6                 Netherlands                   Western Europe   
6                      Canada                              NaN   
7                 New Zealand        Australia and New Zealand   
7                      Sweden                   Western Europe   
7                 New Zealand                              NaN   
8                   Australia        Australia and New Zealand   
8                 New Zealand        Australia and New Zealand   
8                      Sweden                              NaN   
9                      Sweden                   Western Europe   
9                   Australia        Australia and New Zealand   
9                   Australia                              NaN   
..                        ...                              ...   
146                     Yemen  Middle East and Northern Africa   
147                   Liberia                              NaN   
147                Madagascar               Sub-Saharan Africa   
147  Central African Republic               Sub-Saharan Africa   
148                  Tanzania               Sub-Saharan Africa   
148                      Chad               Sub-Saharan Africa   
148                    Guinea                              NaN   
149                   Liberia               Sub-Saharan Africa   
149                    Guinea               Sub-Saharan Africa   
149                      Togo                              NaN   
150               Ivory Coast               Sub-Saharan Africa   
150                    Rwanda                              NaN   
150                    Guinea               Sub-Saharan Africa   
151              Burkina Faso               Sub-Saharan Africa   
151                    Rwanda               Sub-Saharan Africa   
151                     Syria                              NaN   
152               Afghanistan                    Southern Asia   
152                     Benin               Sub-Saharan Africa   
152                  Tanzania                              NaN   
153               Afghanistan                    Southern Asia   
153                   Burundi                              NaN   
153                    Rwanda               Sub-Saharan Africa   
154  Central African Republic                              NaN   
154                      Togo               Sub-Saharan Africa   
154                     Benin               Sub-Saharan Africa   
155                     Syria  Middle East and Northern Africa   
155                     Syria  Middle East and Northern Africa   
156                   Burundi               Sub-Saharan Africa   
156                   Burundi               Sub-Saharan Africa

Y hasta aquí puedo leer, podría con estos datos hacer un ranking de los países que más se han repetido en los 3 primeros puestos por ejemplo y ver la evolución de la renta per caṕita anual para ver si tiene que ver con el scoring por poner un ejemplo. También me hubiera gustado hacer bins, pero he llegado hasta aquí